### PR TITLE
[WIP] Add Wayland backend to generic_unix

### PR DESCRIPTION
**Pull request by ddevault**
Depends on https://github.com/godotengine/godot/pull/27266.

I started with @toger5's work on #23426 and cleaned things up a bit, then picked up the mantle of implementing more of the OS API for Wayland. This is still a WIP, but feedback is welcome.

Major todo items:

- [x] Improved HiDPI support
- [x] Video driver selection
- [x] Update build system
- [x] Choose the most suitable backend on startup in main.cpp
- [x] Compile Wayland protocols with scons
- [x] Re-evaluate EGL context creation
- [x] Default Godot cursor images
- [x] Mouse visible/hidden
- [x] Pointer constraints (for locking the mouse to the window)
- [ ] Clipboard support
- [x] Keep screen on (idle inhibit)

Future work:

- [ ] xdg-decoration support (borderless & bordered windows)
- [ ] Adjust editor DPI auto-selection as the window moves between outputs of various DPIs
- [ ] Key repeat
- [ ] Pointer frames
- [ ] Virtual keyboard (OSK) rigging
- [ ] Custom cursor images

Decorations is probably going to be tough, as it will require Godot to render window decorations itself. The only Wayland compositor which requires this, iiuc, is GNOME. We can probably get away with fixing it separately from the main patch.

### CODE REVIEW

File: platform/generic_unix/wayland/context_egl_wayland.cpp [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	// 	contextAttribs[0] = EGL_CONTEXT_MINOR_VERSION;
+	// 	contextAttribs[1] = 3;
+	// 	contextAttribs[4] = EGL_NONE;
+	// } else {
+	// 	contextAttribs[0] = EGL_CONTEXT_MAJOR_VERSION;
+	// 	contextAttribs[1] = 3;
+	// 	contextAttribs[0] = EGL_CONTEXT_MINOR_VERSION;
+	// 	contextAttribs[1] = 3;
+	// 	contextAttribs[4] = EGL_NONE;
+	// }
```
Comment (toger5) [inline platform/generic_unix/wayland/context_egl_wayland.cpp]: Could you also have a look at this section? I was just trying around what works.

**Comment by toger5** [comment]
About client side decorations. I think that is a nice to have thing for Godot in general. I would not do it in this specific PR. But having the option to run the editor with client side decorations is a great option also for windows/macos (gives a tiny amount of additional screen space). It also does not require much. just adding a close + minimize + fullscreenbutton in the toplft/right corner and maybe optionally choose the style based on the os.

File: platform/generic_unix/wayland/context_egl_wayland.cpp (hunk 1 above)
Comment (ddevault) [reply platform/generic_unix/wayland/context_egl_wayland.cpp]: Sure thing.

File: platform/generic_unix/wayland/os_wayland.cpp [hunk 2]
```diff
@@ -487,8 +487,14 @@ int OS_Wayland::get_screen_dpi(int p_screen) const {
 	return outputs[p_screen]->scale * 96;
 }
 
+OS_GenericUnix *OS_Wayland::detect() {
+	if (!getenv("WAYLAND_DISPLAY") && !getenv("WAYLAND_SOCKET")) {
```
Comment (ddevault) [inline platform/generic_unix/wayland/os_wayland.cpp]: Strictly speaking, a better solution would be to actually attempt to establish a connection with the display (ditto for X11), but that would move this task much earlier in the startup process than it was before. Open to feedback on this approach.

File: platform/generic_unix/wayland/context_egl_wayland.cpp (hunk 1 above)
Comment (ddevault) [reply platform/generic_unix/wayland/context_egl_wayland.cpp]: Took at look at this just now. We don't need to specify these things, and it's better to use the defaults here. However, while studying this code I found that we can use a more conservative `EGL_CONTEXT_CLIENT_VERSION`, and updated the EGL config accordingly.

**Comment by ddevault** [comment]
I moved the code generation into the build system and pulled in my scons module for Wayland protocols as a git subtree:

https://git.sr.ht/~sircmpwn/scons-wayland-scanner

Not sure if this is entirely couth, but it fixes CI issues where checked-in generated protocol files were causing linter errors for code style.

For those unfamiliar with a git subtree, it pulls in commits from an external repository and rewrites them under a subdirectory in this repository. The files are actually checked into the repo and don't require a second step or reliable access to a second git remote to download the module on git clone. The subtree can be updated from upstream in the future with a command like `git subtree pull -P platform/generic_unix/tools/wayland-scanner https://git.sr.ht/~sircmpwn/scons-wayland-scanner master`.

**Comment by toger5** [comment]
Just to make testing for others quicker. 
These additional libs are needed on fedora:
**mesa-libEGL-devel**
**egl-wayland-devel**
**libxkbcommon-devel**
**wayland-devel**

(With these it works.. maybe one of them is not needed...)

**Comment by toger5** [comment]
I have issues when running:
```
wl_registry@2: error 0: invalid version for global wl_output (4): have 2, wanted 3
WARNING: _get: Property not found: rendering/quality/driver/driver_fallback
   At: core/project_settings.cpp:193.
ERROR: is_viable: Error initializing GLAD
   At: drivers/gles3/rasterizer_gles3.cpp:142.
WARNING: _get: Property not found: rendering/quality/driver/driver_fallback
   At: core/project_settings.cpp:193.
**ERROR** OS_Unix::execute - Could not create child process while executing: xmessage
```
dont have that much addional information. (fedora 29)

**Comment by toger5** [comment]
what is the status on this pr?

**Comment by ddevault** [comment]
Been radio silence from upstream, I dunno.

**Comment by qarmin** [comment]
Will work on this PR still continue?
If not, would it be possible to rebase pr with already implemented functions to the master branch?
In the meantime a significant part of the code was moved to displayerver class and x11 was changed to linuxbsd (of course where it did not refer to xorg).

**Comment by akien-mga** [comment]
This would likely be very difficult to rebase given the extent of the changes in the `master` branch for the OS / DisplayServer split. (Which is also why this and other related PRs have not been reviewed for merging, as they were all pending on this DisplayServer split to happen for 4.0.)

The better approach here would likely be to restart from scratch, copying code from this PR to the relevant new files (`platform/linuxbsd/display_server_wayland.cpp`, etc.).

I'd personally favor a focus on first fixing new issues that we're having with X11 since the DisplayServer split (as some might be in DisplayServer or Window code and would impact Wayland too), but re-starting the work on Wayland now would be appropriate too.

**Comment by aaronfranke** [comment]
Closing this and adding "salvageable" due to the above comment.

Godot is still a fairly buggy from the DisplayServer split, it might be advisable to wait a few months to work on this to avoid chasing bugs that aren't your fault.

**Comment by akien-mga** [comment]
See https://github.com/godotengine/godot-proposals/issues/990 for the new reference proposal (please help flesh it out with technical details, I just opened it as a base for further discussion about the implementation).

**Pull request closed without merging**